core/src/main/java/org/apache/iceberg/SnapshotOperations.java

```diff
@@ -0,0 +1,263 @@
+    validateSnapshots(lastSequenceNumber);
+    validateRefs(currentSnapshotId);
+  }
+
+  private void validateSnapshots(long lastSequenceNumber) {
+    for (Snapshot snap : snapshots) {
+      ValidationException.check(
+          snap.sequenceNumber() <= lastSequenceNumber,
+          "Invalid snapshot with sequence number %s greater than last sequence number %s",
+          snap.sequenceNumber(),
+          lastSequenceNumber);
+    }
+  }
+
+  private void validateRefs(long currentSnapshotId) {
+    for (SnapshotRef ref : refs.values()) {
+      Preconditions.checkArgument(
+          snapshotsById.containsKey(ref.snapshotId()),
+          "Snapshot for reference %s does not exist in the existing snapshots list",
+          ref);
+    }
+
+    SnapshotRef main = refs.get(SnapshotRef.MAIN_BRANCH);
+    if (currentSnapshotId != -1) {
+      Preconditions.checkArgument(
+          main == null || currentSnapshotId == main.snapshotId(),
+          "Current snapshot ID does not match main branch (%s != %s)",
+          currentSnapshotId,
+          main != null ? main.snapshotId() : null);
+    } else {
+      Preconditions.checkArgument(
+          main == null, "Current snapshot is not set, but main branch exists: %s", main);
+    }
+  }
+
+  private static Map<Long, Snapshot> indexSnapshotsById(List<Snapshot> snapshots) {
+    return SerializableMap.copyOf(
+        snapshots.stream().collect(Collectors.toMap(Snapshot::snapshotId, Function.identity())));
+  }
+
+  public static Builder buildFrom(SnapshotOperations base) {
+    return new Builder(base);
+  }
+
+  public static Builder buildFromEmpty() {
+    return new Builder();
+  }
+
+  public static SnapshotOperations empty() {
+    return SnapshotOperations.buildFromEmpty().build();
+  }
+
+  static class Builder {
+    private List<Snapshot> snapshots;
+    private SerializableSupplier<List<Snapshot>> snapshotsSupplier;
+    private final Map<Long, Snapshot> snapshotsById;
+    private Map<String, SnapshotRef> refs;
+    private SerializableSupplier<Map<String, SnapshotRef>> refsSupplier;
+
+    Builder() {
+      this.snapshots = Lists.newArrayList();
+      this.refs = Maps.newHashMap();
+
+      this.snapshotsById = Maps.newHashMap();
+    }
+
+    Builder(SnapshotOperations base) {
+      this.snapshots = Lists.newArrayList(base.snapshots);
+      this.snapshotsSupplier = base.snapshotsSupplier;
+      this.refs = Maps.newHashMap(base.refs);
+      this.refsSupplier = base.refsSupplier;
+
+      this.snapshotsById = indexSnapshotsById(snapshots);
+    }
+
+    Builder snapshots(List<Snapshot> snapshots1) {
+      this.snapshots = snapshots1;
+      return this;
+    }
+
+    Builder snapshotsSupplier(SerializableSupplier<List<Snapshot>> snapshotsSupplier1) {
+      this.snapshotsSupplier = snapshotsSupplier1;
+      return this;
+    }
+
+    Snapshot snapshot(Long id) {
+      return snapshotsById.get(id);
+    }
+
+    Builder add(Snapshot snapshot) {
```

Comment:
Since there are two types being managed by this (snapshots and refs) I think it would be more clear to name all the snapshot methods like `addSnapshot` similar to `addRef`. I was reading the `TableMetadata` code and it took me a second to realize what was happening.